presto-kudu/src/main/java/io/prestosql/plugin/kudu/KuduSplitManager.java

```diff
@@ -59,4 +79,56 @@ public ConnectorSplitSource getSplits(
                 throw new IllegalArgumentException("Unknown splitSchedulingStrategy: " + splitSchedulingStrategy);
         }
     }
+
+    private static CompletableFuture<?> whenCompleted(DynamicFilter dynamicFilter)
+    {
+        if (dynamicFilter.isAwaitable()) {
+            return dynamicFilter.isBlocked().thenCompose(o -> whenCompleted(dynamicFilter));
+        }
+        else {
+            return DynamicFilter.NOT_BLOCKED;
+        }
+    }
+
+    private static class KuduDynamicFilteringSplitSource
+            implements ConnectorSplitSource
+    {
+        private final CompletableFuture<ConnectorSplitSource> future;
+
+        private KuduDynamicFilteringSplitSource(CompletableFuture<ConnectorSplitSource> future)
+        {
+            this.future = future;
+        }
+
+        @Override
+        public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHandle partitionHandle, int maxSize)
+        {
+            return future.thenCompose(splitSource -> splitSource.getNextBatch(partitionHandle, maxSize));
+        }
+
+        @Override
+        public void close()
+        {
+            if (!future.cancel(true)) {
```

Comment:
this still doesn't help with race condition, e.g: other thread might be just after obtaining split source via `getSplitSource(table, splitSchedulingStrategy, dynamicFilter)` but before exiting `.thenApply(result -> getSplitSource(table, splitSchedulingStrategy, dynamicFilter));`. In that case cancellation would be successful, but split source is not closed.

Reply:
updated